src/java.base/share/classes/java/text/MessageFormat.java

```diff
@@ -1651,6 +1819,90 @@ private static void copyAndFixQuotes(String source, int start, int end,
         }
     }
 
+    // Corresponding to the FormatType pattern
+    private enum FormatType {
+        NUMBER("number"),
+        DATE("date"),
+        TIME("time"),
+        DTF_DATE("dtf_date"),
+        DTF_TIME("dtf_time"),
+        DTF_DATETIME("dtf_datetime"),
+        CHOICE("choice"),
+        LIST("list"),
+
+        // Pre-defined DateTimeFormatter types
+        BASIC_ISO_DATE("BASIC_ISO_DATE"),
+        ISO_LOCAL_DATE("ISO_LOCAL_DATE"),
+        ISO_OFFSET_DATE ("ISO_OFFSET_DATE"),
+        ISO_DATE("ISO_DATE"),
+        ISO_LOCAL_TIME("ISO_LOCAL_TIME"),
+        ISO_OFFSET_TIME("ISO_OFFSET_TIME"),
+        ISO_TIME("ISO_TIME"),
+        ISO_LOCAL_DATE_TIME("ISO_LOCAL_DATE_TIME"),
+        ISO_OFFSET_DATE_TIME("ISO_OFFSET_DATE_TIME"),
+        ISO_ZONED_DATE_TIME("ISO_ZONED_DATE_TIME"),
+        ISO_DATE_TIME("ISO_DATE_TIME"),
+        ISO_ORDINAL_DATE("ISO_ORDINAL_DATE"),
+        ISO_WEEK_DATE("ISO_WEEK_DATE"),
+        ISO_INSTANT("ISO_INSTANT"),
+        RFC_1123_DATE_TIME("RFC_1123_DATE_TIME");
+
+        private final String text;
+
+        FormatType(String text){
+            this.text = text;
+        }
+
+        // This method returns a FormatType that matches the passed String.
+        // If no matching FormatType is found, an IllegalArgumentException is thrown.
+        private static FormatType fromString(String text) {
+            for (FormatType other : values()) {
+                // Also check trimmed lower case for historical reasons
+                if (text.trim().toLowerCase(Locale.ROOT)
+                        .equals(other.text.trim().toLowerCase(Locale.ROOT))) {
+                    return other;
+                }
+            }
+            throw new IllegalArgumentException();
+        }
+    }
```

Comment:
Can this be replaced with `Enum.valueOf(String)`, after input is normalized?

Reply:
Good point, changed and also removed the `.text` field from the `FormatType` enum, as the enum name is 1 to 1 with the input String.